### PR TITLE
TED-LIUM Release2 corpus

### PR DESCRIPTION
**Pull request by ZhouW321**

### CODE REVIEW

**Review by christophmluscher — APPROVED** [review]

**Review by JackTemaki — APPROVED** [review]

File: datasets/tedlium2.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+                    speaker.name = spk_name
+                    if "female" in gender:
+                        speaker.attribs["gender"] = "female"
+                    elif "male" in gender:
+                        speaker.attribs["gender"] = "male"
+                    c.add_speaker(speaker)
+
+                if rec_name != last_rec_name:
+                    if recording:
+                        c.recordings.append(recording)
```
Comment (michelwi) [inline datasets/tedlium2.py:182]:
```suggestion
                        c.add_recording(recording)
```
this function also registers `c` as `recording.corpus`

File: datasets/tedlium2.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+                    seg_id = 1
+
+                segment = corpus.Segment()
+                segment.name = str(seg_id)
+                segment.start = float(start)
+                segment.end = float(end)
+                segment.speaker_name = spk_name
+                segment.orth = text
+
+                recording.segments.append(segment)
```
Comment (michelwi) [inline datasets/tedlium2.py:196]:
```suggestion
                recording.add_segment(segment)
```
again use the appropriate adding function for additional checks and registrations

File: datasets/tedlium2.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                seg_id += 1
+
+            if recording:
+                c.recordings.append(recording)
```
Comment (michelwi) [inline datasets/tedlium2.py:200]:
```suggestion
            # add final recording
            c.add_recording(recording)
```
I think there is no need for the `if`, if `recording is None` at this point, there is more trouble and we deserve to fail xD

File: datasets/tedlium2.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                if "ignore_time_segment_in_scoring" in line:
+                    continue
+                line = line.replace("<F0_M>", "<o,f0,male>")
+                line = line.replace("<F0_F>", "<o,f0,female>")
+
+                tokens = line.split()
+                assert len(tokens) >= 7, line
+                recName, channel, spkName, start, end, gender = tokens[:6]
+                text = " ".join(tokens[6:])
+                data.append([recName, channel, spkName, start, end, gender, text])
```
Comment (michelwi) [inline datasets/tedlium2.py:221]:
```suggestion
                rec_name, channel, spk_name, start, end, gender = tokens[:6]
                text = " ".join(tokens[6:])
                data.append([rec_name, channel, spk_name, start, end, gender, text])
```
again prefer snake_case

**Review by michelwi — CHANGES_REQUESTED** [review]
I do not want to strictly require these changes, but please lets discuss before merging.

File: datasets/tedlium2.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    Download full TED-LIUM Release 2 corpus from
+    https://projets-lium.univ-lemans.fr/wp-content/uploads/corpus/TED-LIUM/
+    (all train/dev/test/LM/dictionary data included)
+    """
+
+    def __init__(self):
+        self.out_corpus_folders = dict(
+            (corpus_key, self.output_path(corpus_key))
+            for corpus_key in ["train", "dev", "test"]
+        )
```
[the remark this answers is not in the document]
Comment (ZhouW321) [reply datasets/tedlium2.py]: this was just copying from recognition/AdvancedTreeSearchJob
but ok, changed

File: datasets/tedlium2.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        subprocess.check_call(["tar", "-zxvf", "TEDLIUM_release2.tar.gz"])
+        for corpus_key in ["train", "dev", "test"]:
+            shutil.move(
+                "TEDLIUM_release2/%s" % corpus_key,
+                self.out_corpus_folders[corpus_key].get_path(),
+            )
+        shutil.move("TEDLIUM_release2/LM", self.out_lm_folder.get_path())
+
+    def process_dict(self):
```
[the remark this answers is not in the document]
Comment (ZhouW321) [reply datasets/tedlium2.py:49]: added

File: datasets/tedlium2.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+            ";;",
+        ]
+        for corpus_key in ["train", "dev", "test"]:
+            f = open("%s.stm" % corpus_key, "w")
+            f.write("\n".join(header) + "\n")
+
+            stm_folder = os.path.join(self.corpus_folders[corpus_key].get_path(), "stm")
+            for stm_file in sorted(os.listdir(stm_folder)):
+                if not stm_file.endswith(".stm"):
+                    continue
```
[the remark this answers is not in the document]
Comment (ZhouW321) [reply datasets/tedlium2.py:129]: this should be clear that there're multiple stm files

File: datasets/tedlium2.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+                preEnd = float(preSeg[4])
+            else:
+                preEnd = 0.0
+            start = max(start - startPad, preEnd)
+            # end padding (next seg not yet padded and start padding is more important)
+            nextStart = end + endPad
+            if nextSeg is not None and seg[0] == nextSeg[0]:
+                nextStart = max(float(nextSeg[3]) - startPad, end)
+            end = min(end + endPad, nextStart)
+            return "%.2f" % (start), "%.2f" % (end)
```
[the remark this answers is not in the document]
Comment (ZhouW321) [reply datasets/tedlium2.py]: sure, changed

File: datasets/tedlium2.py (hunk 1 above)
Comment (ZhouW321) [reply datasets/tedlium2.py]: this was also just copied from datasets/librispeech
changed now

File: datasets/tedlium2.py (hunk 3 above)
Comment (ZhouW321) [reply datasets/tedlium2.py]: ok, changed

File: datasets/tedlium2.py (hunk 4 above)
Comment (ZhouW321) [reply datasets/tedlium2.py]: yep (again forgotten-changed copy from older pipeline)

**Review by michelwi — APPROVED** [review]

**Pull request merged**